### PR TITLE
[noderesourcetopology] update NRT when attributes change

### LINKED ISSUE #621: Kubelet configuration change without immediate cache updates
### Area

- [X] Scheduler
- [ ] Controller
- [ ] Helm Chart
- [ ] Documents

### Other components

_No response_

### What happened?

func (ov *OverReserve) GetCachedNRTCopy(nodeName string, pod *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, bool) {
	ov.lock.Lock()
	defer ov.lock.Unlock()
	if ov.nodesWithForeignPods.IsSet(nodeName) {
		return nil, false
	}

	nrt := ov.nrts.GetNRTCopyByNodeName(nodeName)
	if nrt == nil {
		return nil, true
	}
	nodeAssumedResources, ok := ov.assumedResources[nodeName]
	if !ok {
		return nrt, true
	}

	klog.V(6).InfoS("nrtcache NRT", "logID", klog.KObj(pod), "vanilla", stringify.NodeResourceTopologyResources(nrt))
	nodeAssumedResources.UpdateNRT(klog.KObj(pod).String(), nrt)

	klog.V(5).InfoS("nrtcache NRT", "logID", klog.KObj(pod), "updated", stringify.NodeResourceTopologyResources(nrt))
	return nrt, true
}

When obtaining NRTs from the cache during the filter and score stages, it is assumed that there are changes in the topologyManagerPolicy and topologyManagerScope in the computational node/var/lib/kubelet/config.yaml. Then the topologyPolicies in the NRT resources reported by NFD also changed. However, the cache will not be updated immediately, as the condition for cache updates is that NodesMaybeOverReserved can return nodes. At this point, the filter stage will enter the wrong function based on the cache configuration. For example, if there is a computing node node1, and the topologyManagerPolicy in the NRT cache is singleNUMPodLevel, but the topologyManagerScope in config.yaml is modified to Container, the filter stage will still enter the singleNUMPodLevelHandler function for operation instead of the singleNUMAContainerLevelHandler.

### What did you expect to happen?

There should be a way to obtain the latest topologyPolicies when the filter and score stages retrieve cached NRT data, avoiding entering incorrect functions for operations

### How can we reproduce it (as minimally and precisely as possible)?

_No response_

### Anything else we need to know?

_No response_

### Kubernetes version

[root@master1 ~]# kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short. Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.13", GitCommit:"49433308be5b958856b6949df02b716e0a7cf0a3", GitTreeState:"clean", BuildDate:"2023-04-12T12:15:50Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/arm64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.13", GitCommit:"49433308be5b958856b6949df02b716e0a7cf0a3", GitTreeState:"clean", BuildDate:"2023-04-12T12:08:36Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/arm64"}

### Scheduler Plugins version

master

### PR DESCRIPTION
**Pull request by ffromani**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
One of the key assumptions we took when designing the NodeResourceTopology (NRT) plugin is that the kubelet config of the worker node changes VERY rarely, if at all, during the cluster lifetime. As rule of thumb, it was expected to change with a frequency of like once every quarter (3 months) or so, and likely less often. So the event of changing during a scheduling cycle was deemed extremely low.

However,  we fail to notice kubelet configuration changes (the bits we care reported by NRT data) and we update them incidentally when resyncing the cache.
These updates are expected to be rare, but still failing to noticing them is a different issue because it can lead to unnecessarily bad scheduling decision.

Simply put, failing to noticing these updates is a gap, which we address with this PR

#### Which issue(s) this PR fixes:
Fixes #621 

#### Special notes for your reviewer:
still WIP, more test ongoing and upcoming

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by ffromani** [comment]
the integration test is pretty much done but thare are known issues I'm investigating.

File: pkg/noderesourcetopology/cache/overreserve.go [hunk 1]
```diff
@@ -181,14 +216,22 @@ func (ov *OverReserve) NodesMaybeOverReserved(lh logr.Logger) []string {
...
 		nodes.Incr(node)
 	}
 
+	// always use local copies
+	configChangeNodes := ov.nodesWithAttrUpdate.Clone()
+	configChangeCount := configChangeNodes.Len()
+
 	if nodes.Len() > 0 {
-		lh.V(4).Info("found dirty nodes", "foreign", foreignCount, "discarded", nodes.Len()-foreignCount, "total", nodes.Len())
+		lh.Info("found dirty nodes", "foreign", foreignCount, "discarded", overreservedCount, "configChange", configChangeCount, "total", nodes.Len())
```
Comment (Tal-or) [inline pkg/noderesourcetopology/cache/overreserve.go:229]: Why did we changed from `V(4)` to `Info`

**Review by Tal-or — COMMENTED** [review]
This new functionality is opt-in in order to support backward compatibility.
But IIUC, the current behavior might causes wrong scheduling decision,  would it be wise to keep the default ("buggy") behavior as an option at all?

File: pkg/noderesourcetopology/cache/overreserve.go (hunk 1 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/overreserve.go:229]: debug change which I forgot to revert, fixing

**Comment by ffromani** [comment]
> Thank you for working on this.
> 
> This new functionality is opt-in in order to support backward compatibility. But IIUC, the current behavior might causes wrong scheduling decision, would it be wise to keep the default ("buggy") behavior as an option at all?

That's a good point. We add a extra watch, but previously the behavior was arguably buggy. Let me switch the defaults.

**Comment by Tal-or** [comment]
/hold
/lgtm 

Hold letting other reviews time to chime in

**Comment by ffromani** [comment]
@PiotrProkop howdy! Do you have any comments by any chance?

**Comment by PiotrProkop** [comment]
@ffromani sorry I was on vacations. Looks good to me!
